rust/index/src/fulltext/types.rs

```diff
@@ -407,101 +408,109 @@ impl<'me> FullTextIndexReader<'me> {
-            .posting_lists_blockfile_reader
-            .get_by_prefix(first_token)
+
+        // Retrieve posting lists for each token.
+        let posting_lists = futures::stream::iter(tokens)
+            .then(|token| async {
+                let positional_posting_list = self
+                    .posting_lists_blockfile_reader
+                    .get_by_prefix(token.text.as_str())
+                    .await?;
+                if positional_posting_list.is_empty() {
+                    return Err(FullTextIndexError::EmptyValueInPositionalPostingList);
+                }
+                Ok(positional_posting_list)
+            })
+            .collect::<Vec<_>>()
             .await
-            .unwrap();
-        for (_, doc_id, positions) in first_token_positional_posting_list.iter() {
-            candidates.insert(*doc_id, positions.to_vec());
-        }
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let num_tokens = posting_lists.len();
+        let mut pointers = vec![0; num_tokens];
+        let mut results = Vec::new();
+
+        loop {
+            // Get current doc_ids from each posting list (aka for each token).
+            let current_doc_ids: Vec<Option<u32>> = posting_lists
+                .iter()
+                .enumerate()
+                .map(|(i, posting_list)| {
+                    if pointers[i] < posting_list.len() {
+                        Some(posting_list[pointers[i]].1)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
 
-        // Iterate through the rest of the tokens, intersecting the posting lists with the candidates.
-        let mut token_offset = 0;
-        for (token, _) in token_frequencies[1..].iter() {
-            token_offset += 1;
-            let positional_posting_list = self
-                .posting_lists_blockfile_reader
-                .get_by_prefix(token.as_str())
-                .await
-                .unwrap();
-            // TODO once we sort by frequency, we need to find the token position
-            // here, taking into account which positions for repeats of the same
-            // token have already been used up.
-            // let token_offset = tokens
-            //     .iter()
-            //     .find(|t| t.text == *token)
-            //     .unwrap()
-            //     .offset_from as i32;
-            let mut new_candidates: HashMap<u32, Vec<u32>> = HashMap::new();
-            for (doc_id, positions) in candidates.iter() {
-                let mut new_positions = vec![];
-                for position in positions {
-                    if let Some(positions) =
-                        // Find the positional posting list with second item = to doc_id
-                        positional_posting_list
-                            .iter()
-                            .find(|x| x.1 == *doc_id)
```

Comment:
previously, the majority of time was spent on this line according to my profiler